doc: add acrn-dm parameter emul explanation (#6445)
projectacrn/acrn-hypervisor: doc/user-guides/acrn-dm-parameters.rst

       ``hd`` (harddisk) or ``cd`` (CD-ROM). ``<filepath>`` is the path for the
       backend file and could be a partition name or a regular file, e.g.,
       ``ahci,hd:/dev/sda``.

   * - ``ahci-hd``
     - This is an alias for ``ahci``.

   * - ``ahci-cd``
     - Advanced Host Controller Interface used to connect with AT Attachment
       Packet Interface device (for CD-ROM emulation). ``ahci-cd`` has the same
       parameters with ``ahci``.

   * - ``amd_hostbridge``
     - Virtualized PCI AMD hostbridge

   * - ``hostbridge``
     - Virtualized PCI hostbridge, a hardware bridge between the CPU's
       high-speed system local bus and the Peripheral Component Interconnect
       (PCI) bus.

   * - ``virtio-blk``
     - Virtio block type device, a string could be appended to specify the disk
       image, e.g., ``virtio-blk,/root/test.img``.  You can also could use
       ``nodisk`` to create a virtio-blk device with a dummy backend.

   * - ``virtio-coreu``
     - Used for Protected Audio Visual Path (PAVP) session management to provide
       a User VM with Protected Audio Visual Path service.

   * - ``virtio-input``

>      - Virtio type device to emulate input device. ``evdev`` char device node

Reviewer: What is an evdev?

Author: an virtio input decice ,detailed description is in https://projectacrn.github.io/latest/developer-guides/hld/virtio-input.html?highlight=virtio%20input

       should be appended, e.g., ``-s
       n,virtio-input,/dev/input/eventX[,serial]``. ``serial`` is an optional

>        string used as the Uniq of guest virtio input device.

Reviewer: what is a ``Uniq`` ?

Author: unique identification code for the device (if device has it) ,seen in https://github.com/torvalds/linux/blob/master/include/linux/input.h line 49 
updated in doc

   * - ``virtio-console``
     - Virtio console type device for data input and output.

   * - ``virtio-hyper_dmabuf``
     - Virtio device that allows sharing data buffers between VMs using a
       dmabuf-like interface.

   * - ``virtio-hdcp``
     - Virtio High-bandwidth Digital Content Protection (HDCP) type device. HDCP
       is technology intended to protect unauthorized duplication of high
       definition (HD) video and audio as it travels across connections.

   * - ``virtio-heci``
     - Virtio Host Embedded Controller Interface, parameters should be appended
       with the format ``<bus>:<device>:<function>,d<0~8>``.

   * - ``virtio-i2c``
     - Virtio i2c type device,parameters with format:
       ``<bus>[:<client_addr>[@<node>]][,<bus>[:<client_addr>[@<node>]]``

       * ``<bus>`` specifies the bus number for the native I2C adapter, e.g.,
         ``2`` means ``/dev/i2c-2``.
       * ``<client_addr>`` specifies the address for the native client devices
         such as ``1C`` or  ``2F``.
       * ``@`` specifies the prefix for the ACPI node.
       * ``<node>`` specifies the ACPI node name supported in the

>          ``acpi_node_table[]`` in the source code: only ``cam1``, ``cam2``, and

Reviewer: I don't think this mention of the ``acpi_node_table`` (found in virtio_i2c.c) is needed since there's no further documentation provided there.   If these are the only three names supported, we can just list them and be done, though I'm curious about how additional supported node names are added since they're platform-specific and you mention only one platform ( ``MRB``) has these names defined).

I see there's one additional name in this table you've not mentioned and we should document:  ``default``

Author: which one I did not mention? bus/client_addr/@/node ?

Author: no additional supported node besides the three in APL platform

>          ``hdac`` are supported for MRB and  are platform-specific.

Reviewer: We need a platform name that's more specific than ``MRB``

Author: APL platform updated in doc

   * - ``virtio-gpio``
     - Virtio GPIO type device, parameters format is:
       ``virtio-gpio,<@controller_name{offset|name[=mapping_name]:offset|name[=mapping_name]:...}@controller_name{...}...]>``

       * ``controller_name``: use the command ``ls /sys/bus/gpio/devices`` to
         check the native GPIO controller information.  Usually, the devices
         represent the ``controller_name`` that you can use. You can also use
         the command ``cat /sys/bus/gpio/device/XXX/dev`` to get the device id
         that can be used to match ``/dev/XXX``, and then use ``XXX`` as the

>          ``controller_name``. On MRB and Intel NUC platforms,

Reviewer: Same comment here, We need a platform name that's more specific than ``MRB``, and is this true for all NUCs?

Author: APL platform updated in doc

         ``controller_name`` may be ``gpiochip0``, ``gpiochip1``, ``gpiochip2``,
         and ``gpiochip3``.
       * ``offset|name``: use GPIO offset or its name to locate one native GPIO
         within the GPIO controller.
       * ``mapping_name``: is optional. If you want to use a customized name for
         a FE GPIO, you can set a new name here.

   * - ``virtio-rnd``
     - Virtio random generater type device, with string ``kernel=on`` to
       initialize virtio random generater state.

   * - ``virtio-rpmb``
     - Virtio Replay Protected Memory Block (RPMB) type device, with
       ``physical_rpmb`` to specify RPMB in physical mode,
       otherwise RPMB is in simulated mode.

   * - ``virtio-audio``
     - Virtio audio type device

   * - ``virtio-net``
     - Virtio network type device, parameter should be appended with the format:
       ``virtio-net,<tap_name>,[mac=<XX:XX:XX:XX:XX:XX>]``.  The ``mac`` address
       is optional, ``tap_name`` must include the keyword ``tap``.

   * - ``passthru``
     - Indicates a passthrough device. Optional parameters include:

>        * ``keep_gsi``: keep vGSI for MSI capable ptdev.

Reviewer: is ptdev an abbreviation for passthrough device?  We should write that out if it is.

Author: done

       * ``no_reset``: passthrough PCI devices are reset by default when
         assigning them to a post-launched VM. This parameter prevents this
         reset for debugging purposes.

>        * ``d3hot_reset``: when launching a  Windows post-launched VM, this

Reviewer: this seems such a weird option for a user to specify.  Can the build tools add this automatically if ``acrn-dm --windows`` is used?

Author: it is only need for passthrough device on windows guest VM

Reviewer 2: But we could use the `is_winvm` flag in our pass-through code to detect that it's a Windows User VM starting and add this automatically, couldn't we?

Author: yes we could use is_winvm to do that and remove d3hot_reset parameter

Reviewer: I suspect we need to leave this ``d2hot_reset`` in the documentation until the code is changed to not require it any more, right?

         parameter should be appended to enable a Windows UEFI ACPI bug fix.
       * ``gpu``: create the dedicated ``igd-lpc`` on ``00:1f.0`` for IGD
         passthrough.
       * ``vmsix_on_msi,<bar_id>``: enables vMSI-X emulation based on MSI
         capability.  The specific virtual bar will be allocated.
       * ``enable_ptm``: enable PCIE precise time measurement mechanism for the
         passthrough device.

   * - ``uart``
     - Emulated PCI UART. Use the parameter with the format
       ``uart,vuart_idx:<0~9>`` to specify hypervisor-emulated PCI vUART index.

   * - ``npk``
     - Intel Trace Hub (known as North Peak or NPK) is a set of hardware blocks
       that produce, switch, and output trace data from multiple hardware,
       firmware, and software sources, used to perform full system debugging.

>        Parameter with the format ``npk,<master_offset>/<master number>``

Reviewer: I guess you need to really know what you're doing to use this since we don't talk about what these master offset and master number terms mean.

Author: you mean we need a HLD of NPK ,as other devices ?

   usage::

      --ssram

.. _emul_config:

Emulated PCI Device Types
****************************

In the acrn-dm ``-s`` or ``--pci_slot`` command line parameter, there is a ``<slot_config`` argument
that contains a string describing the type of emulated PCI device along with optional device-dependent
arguments used for configuration.  Here is a table describing these emulated device types:

.. list-table:: Emulated PCI Device Types
   :header-rows: 1

   * - PCI Device Type string 
     - Description

   * - "xhci"
     - USB controller used to support USB3.0 devices, also supports USB2.0 and USB1.0 devices.

>        Parameter <bus number>-<port number> should be added, physical USB device attached on the bus

Author: updated

>        and port will be detected by User VM and be used as expected, e.g.xhci,1-2,1-2.

Reviewer 2: > @dbkinder , I'm thinking we should also add a note here to indicate that bus and port numbers can be retrieved using `lsusb -t` (or `lsusb --tree`). We mention this above but that's in the context of an example so possibly not the best location. I think it should be moved (or repeated) here. Thoughts?

@dbkinder any thoughts?

   * - "lpc"
     - Low Pin Count (LPC) bus is used to connect low speed devices to the cpu, for example serial port,
       keyboard, or mouse.

   * - "igd-lpc"

>      - Windows graphic driver requires this virtualized LPC device to complete the display function.

Author: "operate" is better ,updated

       selected shm_name.

   * - "ahci"
     - Advanced Host Controller Interface provides advanced features to access Serial ATA (SATA)
       storage devices, such as a hard disk. Parameter <type:><filepath>* should be added, type could be 'hd'(harddisk)
       and 'cd'(CD-ROM). <filepath> is the path for the backend file, could be a partition or a regular file,
       e.g. ahci,hd:/dev/sda.

   * - "ahci-hd"
     - It is an alias for "ahci"

   * - "ahci-cd"
     - Advanced Host Controller Interface used to connect with AT Attachment
       Packet Interface device (for CD-ROM emulation). "ahci-cd" has the same parameters with "ahci".

   * - "amd_hostbridge"
     - Virtualized PCI AMD hostbridge

   * - "hostbridge"
     - Virtualized PCI hostbridge, a hardware bridge between the CPU's high-speed system local bus
       and the Peripheral Component Interconnect (PCI) bus.

   * - "pci_gvt"
     - Intel GVT-g technology provides mediated device passthrough for Intel GPUs (Broadwell and newer).
       It can be used to virtualize the GPU for multiple guest VMs, providing near-native graphics performance.

>    * - "virtio-blk"

Reviewer 2: What about the `,b` (optional) option for this emulated device.

Author: updated

Author: updated

     - Virtio block type device, a string could be appended to specify the disk image,e.g.virtio-blk,/root/test.img,

>        user also could use "nodisk" to create a virtio-blk device with dummy backend.

Reviewer 2: What does that exactly do when you use a dummy backend? What do you see from the User VM; empty disk, just a controller, etc.?

   * - "virtio-coreu"
     - Used for Protected Audio Visual Path (PAVP) session management,with this support,
       user VM is provided Protected Audio Visual Path service.

   * - "virtio-input"
     - Virtio type device to emulate input device, an virtio input decice evdev char device node
       should be appended, e.g. -s n,virtio-input,/dev/input/eventX[,serial], “serial” is
       an optional string. When it is specified, it will be used as the unique identification code
       of guest virtio input device.

   * - "virtio-ipu"

>      - Virtio ipu type device, like camera device.

Reviewer 2: Can you spell out `ipu` (Image Processing Unit?) and capitalize it?

What's not clear to me with this one is what does that do? There seems to be no parameters for it, so what does it look like inside the User VM? Does it look like you have a camera attached? If so, what kind of stream do you get from it?

   * - "virtio-console"
     - Virtio console type device for data input and output.

   * - "virtio-hyper_dmabuf"
     - Virtio device that allows sharing data buffers between VMs using a dmabuf-like interface

   * - "virtio-hdcp"
     - Virtio High-bandwidth Digital Content Protection (HDCP) type device. HDCP
       is technology intended to protect unauthorized duplication of high
       definition (HD) video and audio as it travels across connections.

   * - "virtio-heci"
     - Virtio Host Embedded Controller Interface, parameters should be appended with the format

>        <bus>:<device>:<function>,d<0~8>.

Reviewer 2: I guess these values must be the ones seen from the Service VM?

   * - "virtio-i2c"
     - Virtio i2c type device,parameters with format like
       <bus>[:<client_addr>[@<node>]][:<client_addr>[@<node>]][,<bus>[:<client_addr>[@<node>]][:<client_addr>][@<node>]]
       should be appended. <bus> specifies the bus number for the native I2C adapter, e.g.2 means /dev/i2c-2.
       <client_addr> specifies the address for the native client devices such as 1C, 2F...
       "@" specifies the prefix for the ACPI node.
       <node> specifies the ACPI node name supported in the current code. Currently,only cam1,

>        cam2, and hdac are supported for APL platform. These nodes are platform-specific.

Reviewer 2: Do we only support APL platforms for this parameter?

We don't actively support APL anymore.

What are these `cam1, `cam1` and `hdac` devices?

Author: cam1 cam2 hdac are APL platform APCI node in DSDT table

   * - "virtio-gpio"
     - Virtio GPIO type device, parameters format is like
       virtio-gpio,<@controller_name{offset|name[=mapping_name]:offset|name[=mapping_name]:...}@controller_name{...}...]>
       "controller_name", input ls /sys/bus/gpio/devices to check native gpio controller information. Usually,
       the devices represent the controller_name, you can use it as controller_name directly. You can also
       input cat /sys/bus/gpio/device/XXX/dev to get device id that can be used to match /dev/XXX, then use

>        XXX as the controller_name. On APL platform and Intel NUC platforms, the controller_name are gpiochip0,

Reviewer 2: We do not support APL anymore, what about WHL or TGL?

Author: it is not tested except APL

Reviewer 2: > it is not tested except APL

Do you still expect this _could_ work? Or was it really platform-specific and we should go back to mentionning Apollo Lake (APL) platforms only?

Author: we could just mention it as legacy platform , on which it is working

Reviewer 2: Let's do that!

       gpiochip1, gpiochip2.gpiochip3.
       "offset|name",you can use gpio offset or its name to locate one native gpio within the gpio controller.
       "mapping_name",this is optional, if you want to use a customized name for a FE gpio, you can set a
       new name for a FE virtual gpio.

   * - "virtio-rnd"

>      - Virtio random generater type device, with string "kernal=on" to initialize virtio random generater state.

Reviewer 2: ```suggestion
     - Virtio random generater type device, with string "kernel=on" to initialize virtio random generater state.
```
?

What happens if you don't use that string?

Author: VBSK backend framework is not supported anymore

   * - "virtio-rpmb"
     - Virtio Replay Protected Memory Block (RPMB) type device, with "physical_rpmb" to specify RPMB in physical mode,
       otherwise RPMB is in simulated mode.

   * - "virtio-audio"
     - Virtio audio type device

   * - "virtio-net"
     - Virtio network type device, parameter should be appended with the format
       virtio-net,<tap_name|macvtap_name>,[vhost],[mac=<XX:XX:XX:XX:XX:XX>]
       mac address is optional,"tap_name" needs to include the keyword tap.
       vhost specifies vhost backend, otherwise use VBSU backend.

   * - "passthru"

>      - Passthrough device, parameters below could be appended.

Reviewer 2: You describe the optional flags below but you should also describe the BDF info that needs to be provided, e.g. `-s 2,passthru,0/17/0`

       "keep_gsi", to keep vGSI for MSI capable .
       "no_reset", do reset for passthrough PCI device by default when assigning it to post-launched VM, the parameter
       could enable no_reset for debug only.
       "d3hot_reset", to launch Windows post-launched VM, this parameter should be appended to enable window uefi apci bug fix.
       "gpu", create the dedicated "igd-lpc" on 00:1f.0 for IGD passthrough.
       "vmsix_on_msi", "vmsix_on_msi,<bar_id>" is added to specify the enablement of vMSI-X emulation based on MSI capability.
       The specific virtual bar will be allocated.

>        "enable_ptm", enable PCIE precise time measurement mechnism for the passthrough device.

Reviewer 2: ```suggestion
       "enable_ptm", enable PCIE precise time measurement mechanism for the passthrough device.
```

Author: updated

   * - "uart"
     - Emulated PCI UART, user could use the parameter with the format
       uart,vuart_idx:<0~9> to specify hypervisor emulated pci vuart index.

   * - "npk"
     - Intel Trace Hub (known as North Peak or NPK) is a set of hardware blocks that
       produce, switch, and output trace data from multiple hardware, firmware,
       and software sources, used to perform full system debugging. Parameter with the format

>        npk,<master_offset>/<master number> could be appended to specify the master offset in the physical STMR

Reviewer 2: STMR?

Author: STMR and CSR are bar region for NPK ,CSR is for config ,STMR is for master to output trace

   * - ``igd-lpc``
     - Windows graphic driver requires this virtualized LPC device to operate
       the display function.

   * - ``ivshmem``
     - Inter-VM shared memory (ivshmem) virtualized PCI device used specifically
       for shared memory between VMs. Parameters should be added with the format
       ``ivshmem,<shm_name>,<shm_size>``. ``<shm-name>`` specifies a shared memory
       name, and must be listed in :option:`hv.FEATURES.IVSHMEM.IVSHMEM_REGION`
       as configured using the ACRN configurator tool UI, and needs to start
       with a ``dm:/`` prefix.

   * - ``ahci``
     - Advanced Host Controller Interface provides advanced features to access
       Serial ATA (SATA) storage devices, such as a hard disk. Parameter
       ``<type:><filepath>*`` should be added: ``type`` could be
       ``hd`` (harddisk) or ``cd`` (CD-ROM). ``<filepath>`` is the path for the
       backend file and could be a partition name or a regular file, e.g.,
       ``ahci,hd:/dev/sda``.

   * - ``ahci-hd``
     - This is an alias for ``ahci``.

   * - ``ahci-cd``
     - Advanced Host Controller Interface used to connect with AT Attachment

>        Packet Interface device (for CD-ROM emulation). ``ahci-cd`` has the same

Reviewer 2: ```suggestion
       Packet Interface device (for CD-ROM emulation). ``ahci-cd`` supports the same
```

Author: done

>        parameters with ``ahci``.

Reviewer 2: ```suggestion
       parameters than ``ahci``.
```

   * - ``amd_hostbridge``
     - Virtualized PCI AMD hostbridge

   * - ``hostbridge``
     - Virtualized PCI hostbridge, a hardware bridge between the CPU's
       high-speed system local bus and the Peripheral Component Interconnect
       (PCI) bus.

   * - ``virtio-blk``
     - Virtio block type device, a string could be appended with the format 
       ``virtio-blk,[,b,]<filepath>[,options]``

       * ``[,b,]`` specifies a bootable device and the bootable image location
         when using vsbl as the virtual bootloader. 
       * ``<filepath>`` specifies the path of a file or disk partition. 
         You can also could use ``nodisk`` to create a virtio-blk device with a dummy backend.

>          "nodisk" is used for hot-plugs rootfs after the VM is launched by triggering rescan

Reviewer 2: ```suggestion
         ``nodisk`` is used for hot-plugging a rootfs after the User VM has been launched. It is achieved by triggering a rescan
```

Author: done

>          of virtio-blk device by the guest VM, the empty file will be updated to valid file after rescan.

Reviewer 2: ```suggestion
         of the ``virtio-blk`` device by the User VM. The empty file will be updated to valid file after rescan.
```

       a User VM with Protected Audio Visual Path service.

   * - ``virtio-input``
     - Virtio type device to emulate input device. ``evdev`` char device node
       should be appended, e.g., ``-s
       n,virtio-input,/dev/input/eventX[,serial]``. ``serial`` is an optional
       string used as the unique identification code of guest virtio input device.

   * - ``virtio-ipu``
     - Virtio image processing unit (IPU), it is used to connect
       camera device to system, to convert the raw Bayer image into YUV domain.

   * - ``virtio-console``
     - Virtio console type device for data input and output.

   * - ``virtio-hyper_dmabuf``
     - Virtio device that allows sharing data buffers between VMs using a
       dmabuf-like interface.

   * - ``virtio-hdcp``
     - Virtio High-bandwidth Digital Content Protection (HDCP) type device. HDCP
       is technology intended to protect unauthorized duplication of high
       definition (HD) video and audio as it travels across connections.

   * - ``virtio-heci``
     - Virtio Host Embedded Controller Interface, parameters should be appended
       with the format ``<bus>:<device>:<function>,d<0~8>``. You can find the BDF

>        information from service VM.

Reviewer 2: ```suggestion
       information from the Service VM.
```

       ``<bus>[:<client_addr>[@<node>]][,<bus>[:<client_addr>[@<node>]]``

       * ``<bus>`` specifies the bus number for the native I2C adapter, e.g.,
         ``2`` means ``/dev/i2c-2``.
       * ``<client_addr>`` specifies the address for the native client devices
         such as ``1C`` or  ``2F``.
       * ``@`` specifies the prefix for the ACPI node.
       * ``<node>`` specifies the ACPI node name supported in the
         ``acpi_node_table[]`` in the source code: only ``cam1``, ``cam2``, and
         ``hdac`` are supported for APL platform and  are platform-specific.

   * - ``virtio-gpio``
     - Virtio GPIO type device, parameters format is:
       ``virtio-gpio,<@controller_name{offset|name[=mapping_name]:offset|name[=mapping_name]:...}@controller_name{...}...]>``

       * ``controller_name``: use the command ``ls /sys/bus/gpio/devices`` to
         check the native GPIO controller information.  Usually, the devices
         represent the ``controller_name`` that you can use. You can also use
         the command ``cat /sys/bus/gpio/device/XXX/dev`` to get the device id
         that can be used to match ``/dev/XXX``, and then use ``XXX`` as the

>          ``controller_name``. On APL platform and Intel NUC platforms,

Reviewer 2: Assuming this is generic, I would remove this reference to APL and Intel *NUC* platforms and just say that these are what they appear on Intel platforms.

         ``controller_name`` may be ``gpiochip0``, ``gpiochip1``, ``gpiochip2``,
         and ``gpiochip3``.
       * ``offset|name``: use GPIO offset or its name to locate one native GPIO
         within the GPIO controller.
       * ``mapping_name``: is optional. If you want to use a customized name for
         a FE GPIO, you can set a new name here.

   * - ``virtio-rnd``
     - Virtio random generater type device, with string ``kernel=on`` to

>        select VBSK virtio backend, otherwise use VBSU virtio backend.

Reviewer 2: ```suggestion
       select the VBSK virtio backend. The VBSU virtio backend is used by default.
```

   * - ``virtio-rpmb``
     - Virtio Replay Protected Memory Block (RPMB) type device, with
       ``physical_rpmb`` to specify RPMB in physical mode,
       otherwise RPMB is in simulated mode.

   * - ``virtio-audio``
     - Virtio audio type device

   * - ``virtio-net``
     - Virtio network type device, parameter should be appended with the format:

>        ``virtio-net,<tap_name|macvtap_name>,[vhost],[mac=<XX:XX:XX:XX:XX:XX>]``.

Reviewer 2: ```suggestion
       ``virtio-net,<device_name>,[vhost],[mac=<XX:XX:XX:XX:XX:XX>]``.
```

>        The ``mac`` address is optional, ``tap_name`` must include the keyword ``tap``.

Reviewer 2: ```suggestion
       The ``mac`` address is optional, ``device_name`` is the name of the TAP (or MacVTap) device. It must include the keyword ``tap``.
```

>        ``vhost`` specifies vhost backend, otherwise use VBSU backend.

Reviewer 2: ```suggestion
       ``vhost`` specifies vhost backend, otherwise the VBSU backend is used.
```

         reset for debugging purposes.
       * ``d3hot_reset``: when launching a  Windows post-launched VM, this
         parameter should be appended to enable a Windows UEFI ACPI bug fix.
       * ``gpu``: create the dedicated ``igd-lpc`` on ``00:1f.0`` for IGD
         passthrough.
       * ``vmsix_on_msi,<bar_id>``: enables vMSI-X emulation based on MSI
         capability.  The specific virtual bar will be allocated.
       * ``enable_ptm``: enable PCIE precise time measurement mechanism for the
         passthrough device.

   * - ``uart``
     - Emulated PCI UART. Use the parameter with the format
       ``uart,vuart_idx:<0~9>`` to specify hypervisor-emulated PCI vUART index.

   * - ``npk``
     - Intel Trace Hub (known as North Peak or NPK) is a set of hardware blocks
       that produce, switch, and output trace data from multiple hardware,
       firmware, and software sources, used to perform full system debugging.
       Parameter with the format ``npk,<master_offset>/<master number>``
       specifies the master offset in the physical STMR of the host, and the
       master number owned by the User VM.

   * - ``wdt-i6300esb``

>      - Emulated i6300ESB PCI Watch Dog Timer (WDT) Intel SOC devices used to

Reviewer 2: ```suggestion
     - Emulated i6300ESB PCI Watch Dog Timer (WDT) Intel SOC devices used to
```
```suggestion
     - Emulated i6300ESB PCI Watch Dog Timer (WDT) Intel processors use to
```

>        monitor guest VM.

Reviewer 2: ```suggestion
       monitor User VMs.
```

   usage::

      --ssram

> ----

Reviewer: Delete this line separator and let's add a heading and a description of what this table is:

```
.. _emul_config:

Emulated PCI Device Types
****************************

In the acrn-dm ``-s`` or ``--pci_slot`` command line parameter, there is a ``<slot_config`` argument
that contains a string describing the type of emulated PCI device along with optional device-dependent
arguments used for configuration.  Here is a table describing these emulated device types:
```

Also, let's add a link to this section after the examples of using ``-s`` about line 225 above:
```
For more information about emulated device types, see :ref:`emul_config`.
```

.. list-table:: Emul Items List For PCI Device Type
   :header-rows: 1

   * - Emul Type
     - Description

   * - "xhci"
     - Usb controller used to support USB3.0 devices, provide downward compatibility

   * - "lpc"
     - LPC bus used to connect low speed device to cpu, e.g.serial port

   * - "igd-lpc"
     - Windows graphic driver requires this virtualized LPC device to complete the display function

   * - "ivshmem"
     - Virtualized PCI device used specifically for shared memory between VMs

   * - "ahci"
     - Advanced Host Controller Interface used to connect with SATA storage device

   * - "ahci-cd"
     - Advanced Host Controller Interface used to connect with AT Attachment Packet Interface device

   * - "amd_hostbridge"
     - Virtualized PCI AMD hostbridge

   * - "hostbridge"
     - Virtualized PCI hostbridge

   * - "pci_gvt"
     - Using Intel GVT-g technology

   * - "virtio-blk"
     - Virtio block type device

   * - "virtio-coreu"
     - Used for PAVP session management

   * - "virtio-input"

>      - Virtio type device to emulate input device

Author: like keyboard

>      - Virtio type device to emulate input device

Reviewer: I asked the question assuming you'd put the answer into the documentation not as a comment to this PR.

   * - "virtio-ipu"

>      - Virtio ipu type device

Reviewer: Is this IPU "Intelligence" (for AI) or "Image" (for graphics) processing unit?

Author: like camera device

Reviewer: I asked the question assuming you'd put the answer into the documentation not as a comment to this PR.
